modin/core/dataframe/algebra/groupby.py

```diff
@@ -373,16 +412,196 @@ def try_filter_dict(agg_func, df):
-        partition_dict = {k: v for k, v in agg_func.items() if k in df.columns}
-        return lambda grp: grp.agg(partition_dict)
+
+        grp_has_id_level = df.columns.names[0] == cls.ID_LEVEL_NAME
+        # The 'id' level prevents us from a lookup for the original
+        # partition's columns. So dropping the level.
+        partition_columns = df.columns.droplevel(0) if grp_has_id_level else df.columns
+
+        partition_dict = {k: v for k, v in agg_func.items() if k in partition_columns}
+        return cls._build_callable_for_dict(
+            partition_dict, preserve_aggregation_order, grp_has_id_level
+        )
+
+    @classmethod
+    def _build_callable_for_dict(
+        cls, agg_dict, preserve_aggregation_order=True, grp_has_id_level=False
+    ):
+        """
+        Build callable for an aggregation dictionary.
+
+        Parameters
+        ----------
+        agg_dict : dict
+            Aggregation dictionary.
+        preserve_aggregation_order : bool, default: True
+            Whether to manually restore the order of columns for the result specified
+            by the `agg_func` keys (only makes sense when `agg_func` is a dictionary).
+        grp_has_id_level : bool, default: False
+            Whether the frame we're grouping on has intermediate columns
+            (see ``cls.ID_LEVEL_NAME``).
+
+        Returns
+        -------
+        callable(pandas.core.groupby.DataFrameGroupBy) -> pandas.DataFrame
+        """
+        external_aggs = {}
+        native_aggs = {}
+
+        result_columns = []
+        for col, func, func_name, col_renaming_required in walk_aggregation_dict(
+            agg_dict
+        ):
+            # Filter dictionary
+            dict_to_add = (
+                external_aggs if cls.is_external_aggregation(func) else native_aggs
+            )
+
+            new_value = func if func_name is None else (func_name, func)
+            old_value = dict_to_add.get(col)
+
+            if old_value is not None:
+                assert isinstance(
+                    old_value, list
+                ), f"`{col_renaming_required=}` must be `True` when multiple aggregations per columns is specified."
+                old_value.append(new_value)
+            else:
+                dict_to_add[col] = [new_value] if col_renaming_required else new_value
+
+            # Construct resulting columns
+            if col_renaming_required:
+                func_name = str(func) if func_name is None else func_name
+                result_columns.append(
+                    (*(col if isinstance(col, tuple) else (col,)), func_name)
+                )
+            else:
+                result_columns.append(col)
+
+        result_columns = pandas.Index(result_columns)
+
+        def aggregate_on_dict(grp_obj, *args, **kwargs):
+            """Aggregate the passed groupby object."""
+            if len(native_aggs) == 0:
+                native_agg_res = None
+            elif grp_has_id_level:
```

Comment:
this means that if no `native_aggs` is done, then `grp_has_id_level` is ignored, is that correct? and if so, why?

Reply:
no, this whole if-else statement processes how `grp_has_id_level` affects `native_agg` only, what it means is that:
1. if no `native_aggs` is done then `native_aggs_result` is None
2. if `grp_has_id_level` is specified do special logic for `native_aggs`
3. if none above do regular logic for `native_aggs`

Affecting of `grp_has_id_level` for other objects is [processed later](https://github.com/modin-project/modin/pull/5503/files#diff-ab5817de79d216ae8e94c28983a59f8861e965daab7720653fbddd9472708f05R515).